Source/KSCrash/Source/KSCrash/Recording/Tools/BSG_KSMachHeaders.m

```diff
@@ -16,40 +16,73 @@
     return &bsg_mach_binary_images;
 }
 
+// MARK: - Code synchronisation wrappers
+
+// os_unfair_lock is available from specific OS versions onwards:
+//     https://developer.apple.com/documentation/os/os_unfair_lock
+//
+// It deprecates OSSpinLock:
+//     https://developer.apple.com/library/archive/documentation/System/Conceptual/ManPages_iPhoneOS/man3/spinlock.3.html
+
+void lock() {
+#if defined(__IPHONE_10_0) || defined(__MAC_10_12) || defined(__TVOS_10_0) || defined(__WATCHOS_3_0)
+    os_unfair_lock_lock(&bsg_mach_binary_images_access_lock);
+#else
+    OSSpinLockLock(&bsg_mach_binary_images_access_lock);
+#endif
+}
+
+void unlock() {
+#if defined(__IPHONE_10_0) || defined(__MAC_10_12) || defined(__TVOS_10_0) || defined(__WATCHOS_3_0)
+    os_unfair_lock_unlock(&bsg_mach_binary_images_access_lock);
+#else
+    OSSpinLockUnlock(&bsg_mach_binary_images_access_lock);
+#endif
+}
+
 /**
  * Store a Mach binary image-excapsulating struct in a dynamic array.
  * The array doubles on filling-up.  Typical sizes is expected to be in < 1000 (i.e. 2-3 doublings, at app start-up)
  * This should be called in a threadsafe way; we lock against a simultaneous add and remove.
  */
 void bsg_add_mach_binary_image(BSG_Mach_Binary_Image_Info element) {
     
-    os_unfair_lock_lock(&bsg_mach_binary_images_access_lock);
+    lock();
     
-    // Expand array if necessary
+    // Expand array if necessary.  We're slightly paranoid here.  An OOM is likely to be indicative of bigger problems
+    // but we should still do *our* best not to crash the app.
     if (bsg_mach_binary_images.used == bsg_mach_binary_images.size) {
         bsg_mach_binary_images.size *= 2;
-        bsg_mach_binary_images.contents = (BSG_Mach_Binary_Image_Info *)realloc(bsg_mach_binary_images.contents, bsg_mach_binary_images.size * sizeof(BSG_Mach_Binary_Image_Info));
+        size_t newSize = bsg_mach_binary_images.size * sizeof(BSG_Mach_Binary_Image_Info);
+        void **newData = realloc(bsg_mach_binary_images.contents, newSize);
+        
+        // Exit early if we can't allocate memory
+        if (newData == NULL) {
+            return;
```

Comment:
This could lead to deadlock as the lock taken out earlier hasn't been released. `realloc()` also doesn't return `NULL` according to the header documentation:

>The realloc() function tries to change the size of the allocation pointed
     to by ptr to size, and returns ptr.  If there is not enough room to
     enlarge the memory allocation pointed to by ptr, realloc() creates a new
     allocation, copies as much of the old data pointed to by ptr as will fit
     to the new allocation, frees the old allocation, and returns a pointer to
     the allocated memory.  If ptr is NULL, realloc() is identical to a call
     to malloc() for size bytes.  If size is zero and ptr is not NULL, a new,
     minimum sized object is allocated and the original object is freed.  When
     extending a region allocated with calloc(3), realloc(3) does not guaran-
     tee that the additional memory is also zero-filled.

For maximum safety I think it'd be best to always assign the return value of `realloc()` to `bsg_mach_binary_images.contents` and avoid returning early.

Reply:
Thanks for encouraging me to dig a little deeper.  👍 TIL.  The RETURN VALUES section of [the man page](https://developer.apple.com/library/archive/documentation/System/Conceptual/ManPages_iPhoneOS/man3/realloc.3.html) says:

> RETURN VALUES
>      If successful, calloc(), malloc(), realloc(), reallocf(), valloc(), and aligned_alloc() functions return a pointer to allocated memory.  If there is an error, they return a NULL pointer and set errno to ENOMEM.
> ...
> For realloc(), the input pointer is still valid if reallocation failed....

The [realloc implementation](https://opensource.apple.com/source/Libc/Libc-594.1.4/gen/malloc.c) concurs.

I've added an additional check for success and exit early if the failure conditions outlined above are met.  This way temporary memory starvation only fails to record the mach header info but will keep trying.

Does this approach seem safe enough?